src/databricks/labs/ucx/migration/azure_sp.py

```diff
@@ -0,0 +1,162 @@
+from databricks.sdk.service.workspace import ExportFormat
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class ServicePrincipalMigrationInfo:
+    prefix: str
+    client_id: str
+    principal: str
+    privilege: str
+    # if create access manager and managed identity for this SP
+    replace_with_ac: bool
+    # if a storage credential using this SP already exists
+    already_in_sc: bool
+    # SP's client_secret stored in Databricks secret
+    client_secret: str
+    # if this is a managed identity
+    if_mi: bool
+
+
+class AzureServicePrincipalMigration:
+
+    def __init__(self, ws: WorkspaceClient, csv: str, replace_with_ac: bool):
+        self._ws = ws
+        self._csv = csv if csv is not None else f"/Users/{ws.current_user.me().user_name}/.ucx/azure_storage_account_info.csv"
+        self._use_ac = replace_with_ac if replace_with_ac is not None else False
+
+    def _load_sp_csv(self):
+        """
+        Load SP info from azure_storage_account_info.csv
+        :return:
+        """
+        #TODO: check the download status
+        csv_source = self._ws.workspace.download(self._csv, format=ExportFormat.AUTO)
+        csv_textio = io.TextIOWrapper(csv_source, encoding='utf-8')
+
+        csv_reader = csv.DictReader(csv_textio)
+        first_col = csv_reader.fieldnames[0]
+        for row in csv_reader:
+            if row[first_col].startswith("#"):
+                logger.info(f"Skip migrate Azure Service Principal: {row} to UC storage credential")
+                #TODO: record and persist this skip in a table
+                continue
+            use_ac = False
+            if self._use_ac:
+                use_ac = True
+            elif row[first_col].startswith("-"):
+                use_ac = True
+            sp_migration_info = ServicePrincipalMigrationInfo(
+                prefix = row["prefix"],
+                client_id = row["client_id"],
+                principal = row["principal"],
+                privilege = row["privilege"],
+                replace_with_ac = use_ac,
+                already_in_sc = False,
+                client_secret = "",
+                if_mi = False
+            )
+            yield sp_migration_info
+
+
+    def _check_sp_type(self, sp_list):
+        # check if the sp with given client_id is Service Principal or Managed Identity
+        # May use https://learn.microsoft.com/en-us/graph/api/directoryobject-getbyids?view=graph-rest-1.0&tabs=http
+        # Managed Identity should have "@odata.type": "#microsoft.graph.managedIdentity"
+        # https://learn.microsoft.com/en-us/graph/api/resources/managedidentity?view=graph-rest-beta (warning: beta api)
+        # update if_mi field of the sp in sp_list
+        # TODO: find out if azure databricks access connector with system assigned managed identity
+        # TODO: has role assignment on storage, what ID will be returned by providers/Microsoft.Authorization/roleAssignments
+        return
+
+    def _list_sc(self):
```

Comment:
What is SC? Be more explicit with names

Reply:
it's for storage credential. Just updated the function name.